server/src/main/java/org/opensearch/gateway/GatewayAllocator.java

```diff
@@ -168,6 +201,98 @@ public void allocateUnassigned(
         innerAllocatedUnassigned(allocation, primaryShardAllocator, replicaShardAllocator, shardRouting, unassignedAllocationHandler);
     }
 
+    @Override
+    public void allocateUnassignedBatch(final RoutingAllocation allocation, boolean primary) {
+        // create batches for unassigned shards
+        createBatches(allocation, primary);
+
+        assert primaryBatchShardAllocator != null;
+        assert replicaBatchShardAllocator != null;
+        if (primary) {
+           batchIdToStartedShardBatch.values().forEach(shardsBatch -> primaryBatchShardAllocator.allocateUnassignedBatch(shardsBatch.getBatchedShardRoutings(), allocation));
+        } else {
+            batchIdToStoreShardBatch.values().forEach(batch -> replicaBatchShardAllocator.allocateUnassignedBatch(batch.getBatchedShardRoutings(), allocation));
+        }
+    }
+
+    private void createBatches(RoutingAllocation allocation, boolean primary) {
+        RoutingNodes.UnassignedShards unassigned = allocation.routingNodes().unassigned();
+        // fetch all current batched shards
+        Set<ShardId> currentBatchedShards = primary? startedShardBatchLookup.keySet() : storeShardBatchLookup.keySet();
+        Set<ShardRouting> shardsToBatch = Sets.newHashSet();
+        // add all unassigned shards to the batch if they are not already in a batch
+        unassigned.forEach(shardRouting -> {
+            if ((currentBatchedShards.contains(shardRouting.shardId()) == false) && (shardRouting.primary() == primary)) {
+                assert shardRouting.unassigned();
```

Comment:
Do you want to fail the code here ? or just skip it ?

If the async flow assigns this particular shard, we should still take care of the rest of the shards right.

Reply:
The idea to assert is that the flow should not have reached here. Thats why asserting